tests/Giraffe.Tests/HttpHandlerTests.fs

```diff
@@ -623,48 +609,51 @@ let ``GET "/api/users" returns "users"`` () =
     ctx.Response.Body <- new MemoryStream()
     let expected = "users"
 
-    let result =
+    task {
+        let! result = app next ctx
 
-        app next ctx
-        |> Async.RunSynchronously
-
-    match result with
-    | None -> assertFailf "Result was expected to be %s" expected
-    | Some ctx ->
-        let body = getBody ctx
-        Assert.Equal(expected, body)
+        match result with
+        | None -> assertFailf "Result was expected to be %s" expected
+        | Some ctx ->
+            let body = getBody ctx
+            Assert.Equal(expected, body)
+    }
 
 [<Fact>]
 let ``GET "/api/test" returns "test"`` () =
-    let ctx = Substitute.For<HttpContext>()
-    let app =
-        GET >=> choose [
-            route "/"    >=> text "Hello World"
-            route "/foo" >=> text "bar"
-            subRoute "/api" (
-                choose [
-                    route ""       >=> text "api root"
-                    route "/admin" >=> text "admin"
-                    route "/users" >=> text "users" ] )
-            route "/api/test" >=> text "test"
-            setStatusCode 404 >=> text "Not found" ]
-
-    ctx.Items.Returns (new Dictionary<obj,obj>() :> IDictionary<obj,obj>) |> ignore
-    ctx.Request.Method.ReturnsForAnyArgs "GET" |> ignore
-    ctx.Request.Path.ReturnsForAnyArgs (PathString("/api/test")) |> ignore
-    ctx.Response.Body <- new MemoryStream()
-    let expected = "test"
-
-    let result =
 
-        app next ctx
-        |> Async.RunSynchronously
+    task {
+        let ctx = Substitute.For<HttpContext>()
 
-    match result with
-    | None -> assertFailf "Result was expected to be %s" expected
-    | Some ctx ->
-        let body = getBody ctx
-        Assert.Equal(expected, body)
+        let app =
+            GET >=> choose [
+                route "/"    >=> text "Hello World"
+                //route "/foo" >=> text "bar"
+                subRoute "/api" (
+                    choose [
+                        route ""       >=> text "api root"
+                        route "/admin" >=> text "admin"
+                        route "/users" >=> text "users" ] )
+                route "/foo" >=> text "bar"
+                route "/api/test" >=> text "test" //repositioned
```

Comment:
Apologies, meant to switch back again, I ran into a testing issue where routes after a subroute handler were failing so I switched around the order to test the consistency of the issue. The route order can be put back as was as I fixed the problem.

The issue related to the subroute not stripping back the subpath on a failed match attempt now that the execution path had changed slightly and removed the need to do try/finally in [`handlerWithRootedPath`](https://github.com/dustinmoris/Giraffe/pull/75/files/6b527e6417ea2aefa873b1a3c2ca610a7352acf4?diff=split#diff-d134121409311cd8b419d9dbed9a41bb) function as it can determine success/failure in its own scope now.

Please disregard or change back as does not need to be changed.